audformat/core/utils.py

```diff
@@ -1074,3 +1074,31 @@ def union(
     index = index.drop_duplicates()
 
     return index
+
+
+def iter_index_by_file(
+        index: pd.Index
+) -> (str, typing.Iterator[pd.Index]):
+    r"""Iterate over index by file.
+
+    In each iteration return sub-index for one file.
+
+    Args:
+        index: index object conform to
+            :ref:`table specifications <data-tables:Tables>`
+
+    Returns:
+        iterator in form of (file, sub_index)
+
+    Examples
+        >>> i = segmented_index(['f1'] * 2, [0, 1], [2, 3])
+        >>> print(next(iter_index_by_file(i)))
+        ('f1', MultiIndex([('f1', '0 days 00:00:00', '0 days 00:00:02'),
+            ('f1', '0 days 00:00:01', '0 days 00:00:03')],
+           names=['file', 'start', 'end']))
+    """
+    index = to_segmented_index(index, allow_nat=False)
+    files = index.get_level_values('file').drop_duplicates()
+    for file in files:
+        select = index[index.get_loc(file)]
+        yield file, select
```

Comment:
I'm not so sure if we should really return `file` and `select` as separate values. I know, this is my own implementation, but we know deal with the official API and not just some custom function used in a specific context. What is not so nice here is that `select' is not confirm to `audformat` since it has only `start` and `end`, which we do not officially support, i.e. other utility function will not work. @hagenw what is your take here?

Reply:
I would prefer that it simply returns a segmented index containing a single file and all segments.
For filewise, I'm not sure if it should return a filewise index, or also a segmented index.

Can you provide possible applications of this utility function?

Reply:
As the name of the function and its description only indicates the iteration, I would expect that it returns a filewise index with a single file when feeding it with a filewise index.

Reply:
When I was testing the function offline trying to see what it does, select is actually a segmented index that is conform to `audformat` (see example in the docstring). Isn't that intended ?

Reply:
One is the overlap scenario, of course, when we look for overlaps we only want to consider segments belonging to the same file. Another one might be an evaluation per file:

```python
for index_per_file in iter_index_by_file(segmented_index):
    pred = model.process_index(index_per_file)
    truth = truths.loc(index_per_file)
    ...
```

If we ever allow duplicates in indices again, it would be handy to select all labels that belong to the same files, e.g.

```python
for files in index_per_file(filewise_index):
   labels_belonging_to_the_same_file = df.loc[index_per_file]
```

> For filewise, I'm not sure if it should return a filewise index, or also a segmented index.

My motivation to not convert to a segmented index here is that then we don't have to care about `allow_nat` argument and `num_workers` argument, which we otherwise have to pass on I guess.

Reply:
Mhh, yes probably we should, as we already have it. Otherwise the user has to do 'index.get_level_values('file')[0]` to get it. So I recommend we return the file name and an index, but the index should be conform to `audformat`, i.e. include the `file` level.

Reply:
But according to https://github.com/audeering/audformat/pull/164#discussion_r840581711 this is already the case :)

Reply:
Then if we want to consider all cases, let's return a Filewise index per file if the index object is a filewise index as @hagenw said. If not we add the allow_nat as an argument, and based on that we either apply the changes and return (file, segmented_index) or we return (file, segmented_index) as it is with nat in the indexes

Reply:
Agree, in the example the `file` level is included in the returned index, so that looks fine to me.

To also avoid the `aloow_nat` problem I would then simply return a filewise index for filewise input.

Reply:
> But adding allow_nat as an argument to iter_index_by_file isn't the same as leaving it to the user ? I mean we're just allowing more flexibility in case the user forgets that he needs it :x

I'm in favor of having as less arguments as possible. And especially avoid arguments that only apply in some cases. Cause if the input is filewise, then `allow_nat` has no effect at all. + if we add `allow_nat` then would also need to add `num_workers`.

Reply:
Mhm okay then, let's keep it simple

Reply:
The only argument for adding it, would be if the algorithm requires that we do not have `NaT` values. But this is not the case here.